### ISSUE TITLE
how to use  the  re-ranking?

### ISSUE THREAD

**Issue by Phoebe-star** [huanghoujing/AlignedReID-Re-Production-Pytorch#18]
Python version Re-ranking (Originally from re_ranking)
1.  I have no idea to use it , how to feed data type to use re_ranking
I can not understand it, why is num_query  ,not  feature
code API :
q_g_dist: query-gallery distance matrix, numpy array, shape [num_query, num_gallery]
q_q_dist: query-query distance matrix, numpy array, shape [num_query, num_query]
g_g_dist: gallery-gallery distance matrix, numpy array, shape [num_gallery, num_gallery]

2.   it is for training when training  or   it is for test when test?

thanks , answer can use English  or 中文

**Comment by huanghoujing** [comment]
你好，这是测试时候用的。

测试时候把所有query图片（num_query张图片）的特征提好，所有gallery图片（num_gallery张图片）的特征也提好，这样就可以计算query-gallery之间的一个距离矩阵`q_g_dist`，维度为(num_query, num_gallery)。这样的距离矩阵可以直接调用CMC和mAP函数算分数。

想要re-ranking，那还得计算query-query之间的一个距离矩阵`q_q_dist`，维度为(num_query, num_query)，以及gallery-gallery之间的一个距离矩阵`g_g_dist`，维度为(num_gallery, num_gallery)。这样的几个输入就可以进行re-ranking了。Re-Ranking的结果是一个新的query-gallery距离矩阵`q_g_dist`，这个时候再调用CMC和mAP函数算分数就行。

**Comment by Phoebe-star** [comment]
thanks ,  I have another problems
1. what is the local distance (NxN),
    paper define the shortest path from (1,1) to (H,H), so it is calculating  HxH  times to get the S (total 
    distance)?  or  it is calculating other times ?  why?
    and where is your local distance code? 
![image](https://user-images.githubusercontent.com/29423761/36015235-9614270c-0da8-11e8-844b-8a9d3583b92d.png)

2.the local distance and global distance , how to use hard triplet loss?
   the global distance (NxN) is just to pick up the hard sample?
   if pick up the hard sample of global distance  ,then find the corresponding local distance?
   after,  adding the local distance and global distance to do triplet loss?

3.how is hard sample mining  number ?

**Comment by huanghoujing** [comment]
1. Shortest distance is the accumulation of part-to-part distances, so it should calculate a HxH distance matrix and then accumulate the values along the shortest path.
2. The global triplet loss is a constraint on global feature, local triplet loss on body part features. TriHard (BatchHard) triplet loss requires mining hard positive and negative. The hard triplets for global triplet loss (Here these triplets/samples are denoted by S_global) is searched from the NxN (N is batch size) global distance matrix Dist_global. The hard triplets/samples for local triplet loss can be S_global or searched from the NxN (N is batch size) local distance matrix Dist_local. In this implementation you can try these two types of triplets for local distance with the script argument `local_dist_own_hard_sample`. `local_dist_own_hard_sample=False` means local loss using hard samples S_global, which is the case the authors report the scores in the paper.
3. Number of triplets is the same as batch size.
4. Here `H=7`, so for a pair of images, we first calculate the 7x7 distance matrix between parts of two images, then we calculate shortest distance between these two images from this 7x7 matrix.

**Comment by Phoebe-star** [comment]
in the distance.py

1.how about   m1,  m2, n
def compute_dist(array1, array2, type='euclidean'):
Args:
    array1: numpy array with shape [m1, n]
    array2: numpy array with shape [m2, n]

2.how about m ,n 
def shortest_dist(dist_mat):
Args:
    dist_mat: numpy array, available shape
      1) [m, n]
      2) [m, n, N], N is batch size

3.how about   m , n d
def meta_local_dist(x, y):
  Args:
    x: numpy array, with shape [m, d]
    y: numpy array, with shape [n, d]

4.how about   M,m,d  ,N,n
def parallel_local_dist(x, y):
  Args:
    x: numpy array, with shape [M, m, d]
    y: numpy array, with shape [N, n, d]

**Comment by Phoebe-star** [comment]
Hi
I use the distance.py 
the  function   " def shortest_dist(dist_mat): "
is it return the feature distance, right ? ( it is like doing global feature and comoute the distance of two person images)
then  directly  add  local distance to global distance , to do  triplet loss, right?
if , it  is right ,  I get the local distance , all vale is  10.XXXX ,  (same value)   is it error  
answer can English or 中文  , thanks

**Comment by huanghoujing** [comment]
Closed because I could not explain well.
